### DISCUSSION TITLE
MicroPython Firmware for STM32H562RGT6

### DISCUSSION THREAD

**Discussion by daleka** [micropython/micropython#16159] [Core Development] [answered]
Hi everyone!

I have a WeAct Studio STM32H562RGT6 Core Board, and I noticed there’s no official MicroPython support for this board. However, I see that there is support for the NUCLEO_H563ZI board. Both boards have a similar STM32H5 chip, with the main differences being memory size and connected peripherals.

Is it possible to create a MicroPython firmware for the WeAct STM32H562RGT6 based on the existing support for the NUCLEO_H563ZI? If so, does anyone have instructions or suggestions on how to do this?

For reference, here’s a link to the documentation and board specs for the WeAct STM32H562RGT6: [WeAct Studio STM32H5 64Pin CoreBoard](https://github.com/WeActStudio/WeActStudio.STM32H5_64Pin_CoreBoard).

Any advice or guidance would be greatly appreciated!

**Comment by bidrohini** [comment]
You have to install the required tools, such as arm-none-eabi-gcc for cross-compiling ARM firmware.

**Reply by daleka** [reply]
"Thank you for the reply! Installing arm-none-eabi-gcc and other tools for cross-compiling isn't an issue, but it seems like that alone might not be enough. Since the STM32H562RGT6 board has some differences from the NUCLEO_H563ZI (like memory size and peripheral connections), I think it may require adjusting configuration files, pin definitions, or other specifics to fully support the board.

If you have experience creating or configuring MicroPython for new STM32 boards, I’d appreciate any additional guidance. I suspect modifications to files like mpconfigboard.h, pins.csv, or similar might be necessary, but I’m not entirely sure where to start."

**Reply by AndreoBotelho** [reply]
have you managed to build it already? I'm using it and can share you some of my build files if you want

**Reply by daleka** [reply]
Hi! I haven't been able to build the firmware for this board yet. If you could share anything, I'd appreciate it!

**Comment by AndreoBotelho** [comment, marked as answer]
[WEACT_H562RG.zip](https://github.com/user-attachments/files/17692533/WEACT_H562RG.zip)
uploaded the board definition I'm using, there's some changes I've made to support my needs but you can change the settings you want on mpconfigboard.h

**Reply by daleka** [reply]
"Hi Andre, thank you so much for sharing the board definition and build files! This is incredibly helpful for me, and I truly appreciate your support. My project requires 8 UARTs, and I think I now have a good idea of how to proceed. I'll take a look at the changes you've made and see how I can adapt them to my project. Thanks again for your kindness!"

**Reply by daleka** [reply]
> [WEACT_H562RG.zip](https://github.com/user-attachments/files/17692533/WEACT_H562RG.zip) загрузил определение платы, которое я использую, есть некоторые изменения, которые я сделал для поддержки своих нужд, но вы можете изменить нужные вам настройки в mpconfigboard.h

Hi.

Sorry to bother you, and thank you again for sharing the files for the H562RG — they were incredibly helpful to me!

I just wanted to ask if, by any chance, you have experience working with the WeAct Studio STM32H743VIT6 Core Board? Or perhaps you know if a firmware for this board is available?

I’d greatly appreciate any information!

Thank you!

**Reply by robert-hh** [reply]
THere is  a PR by Matt for this board. I tried the board files and they work. https://github.com/micropython/micropython/pull/12540

**Reply by AndreoBotelho** [reply]
> > [WEACT_H562RG.zip](https://github.com/user-attachments/files/17692533/WEACT_H562RG.zip) загрузил определение платы, которое я использую, есть некоторые изменения, которые я сделал для поддержки своих нужд, но вы можете изменить нужные вам настройки в mpconfigboard.h
> 
> Hi.
> 
> Sorry to bother you, and thank you again for sharing the files for the H562RG — they were incredibly helpful to me!
> 
> I just wanted to ask if, by any chance, you have experience working with the WeAct Studio STM32H743VIT6 Core Board? Or perhaps you know if a firmware for this board is available?
> 
> I’d greatly appreciate any information!
> 
> Thank you!

you can use the VCC_GND_H743VI with some changes to adapt to weact or download the files from the PR. A good way to configure is to use the STM32CubeIDE to help with the pinout, timings and peripherals.

**Reply by mattytrentini** [reply]
I'm sorry for the delay @robert-hh, but I haven't forgotten about those (3x H7 WeAct Studio) boards - I intend to revisit them after PyCon, this weekend.

**Reply by robert-hh** [reply]
@mattytrentini Don't worry, don't hurry.

**Reply by daleka** [reply]
> THere is a PR by Matt for this board. I tried the board files and they work. #12540

Hi Robert,

Thank you for pointing me to the PR (#12540). Unfortunately, I'm having issues with compiling the firmware for the WEACT_STM32H743 board. Here's a summary of the problems I've encountered:

Conflicting definitions:

I ran into redefinition errors for HSE_VALUE, LSE_VALUE, and EXTERNAL_CLOCK_VALUE. These appear to be defined both in the stm32h7xx_hal_conf.h file (specific to the board) and in the stm32h7xx_ll_rcc.h file from the HAL library.
This leads to compilation failure with the -Werror flag, treating these warnings as errors.
Submodule synchronization:

I ensured all submodules were up to date using git submodule update --init --recursive, but the issue persists.
Tried both commit versions:

I attempted to compile with both the commits in the PR: the one marked with a green check and the one with a red cross. Both resulted in similar errors.
Despite these efforts, I couldn't get the build to complete successfully. If you have any suggestions or insights based on your successful experience with these files, I'd greatly appreciate it!

Thank you in advance for your time and assistance!

**Reply by robert-hh** [reply]
I have built  a firmware for a STM32H743 board with a STM32H743VIT6 chip. Attached are the .dfu and .hex files. It works and shows 8MB of flash and about 430kB head space. You can upload it to the board using dfu-util. Set the board to dfu-mode by pushing first the B0 button, then NR and then release both button. Call `dfu.util -l` to check, if DFU capable devices are present, and call `dfu-util --help` for usage. The option for loading firmware to the device is -D. The command accroding to the Makefile is:
`dfu-util -a 0 -d vid:pid -D firmware.dfu -s :leave`
Loading is slow, at least on my set-up.

[stm32H743_firmware.zip](https://github.com/user-attachments/files/18306271/stm32H743_firmware.zip)

Edit: These are my board definition files:

[WEACT_STM32H743.zip](https://github.com/user-attachments/files/18306282/WEACT_STM32H743.zip)

**Reply by daleka** [reply]
> [WEACT_H562RG.zip](https://github.com/user-attachments/files/17692533/WEACT_H562RG.zip) uploaded the board definition I'm using, there's some changes I've made to support my needs but you can change the settings you want on mpconfigboard.h

Hello friends,

First of all, I want to sincerely thank everyone who has replied to me earlier — your help is truly appreciated.

I'm still quite new to hardware configuration, and I must admit it's really difficult for me to get things right on the first try. I apologize if my questions seem too simple or naive.

I'm working with the WeAct Studio STM32H562RGT6 board. Thanks to @AndreoBotelho, I already have a basic MicroPython board configuration that compiles successfully. However, it currently only enables two UARTs, while I critically need at least 5 UARTs for my project (ideally, all 6 supported by the MCU).

Here is what I need working for my project:

At least 5 UARTs (ideally 6),

One SPI interface (used for other peripherals — not for SD card),

SD card support (either via SPI or SDMMC),

USB support, preferably with REPL.

Everything else (I2C, ADC, PWM, etc.) is not needed and can be disabled.

Could someone kindly guide me on how to configure the board to enable these interfaces? I assume I need to edit pins.csv, mpconfigboard.h, mpconfigboard.mk, or similar files — but I'm not sure where to start or how to properly define everything.

I would be extremely grateful for any advice, examples, or working configurations!

**Comment by AndreoBotelho** [comment]
Just add all the uart pinout you need on the mpconfigboard.h file

// UART config
#define MICROPY_HW_UART1_TX         (pin_A9) // SB23: ST-Link
#define MICROPY_HW_UART1_RX         (pin_A10) // SB18: ST-Link
#define MICROPY_HW_UART6_TX         (pin_C6)
#define MICROPY_HW_UART6_RX         (pin_C7)


Em dom., 6 de abr. de 2025 às 14:11, daleka ***@***.***>
escreveu:

> WEACT_H562RG.zip
> <https://github.com/user-attachments/files/17692533/WEACT_H562RG.zip>
> uploaded the board definition I'm using, there's some changes I've made to
> support my needs but you can change the settings you want on mpconfigboard.h
>
> Hello friends,
>
> First of all, I want to sincerely thank everyone who has replied to me
> earlier — your help is truly appreciated.
>
> I'm still quite new to hardware configuration, and I must admit it's
> really difficult for me to get things right on the first try. I apologize
> if my questions seem too simple or naive.
>
> I'm working with the WeAct Studio STM32H562RGT6 board. Thanks to
> @AndreoBotelho <https://github.com/AndreoBotelho>, I already have a basic
> MicroPython board configuration that compiles successfully. However, it
> currently only enables two UARTs, while I critically need at least 5 UARTs
> for my project (ideally, all 6 supported by the MCU).
>
> Here is what I need working for my project:
>
> At least 5 UARTs (ideally 6),
>
> One SPI interface (used for other peripherals — not for SD card),
>
> SD card support (either via SPI or SDMMC),
>
> USB support, preferably with REPL.
>
> Everything else (I2C, ADC, PWM, etc.) is not needed and can be disabled.
>
> Could someone kindly guide me on how to configure the board to enable
> these interfaces? I assume I need to edit pins.csv, mpconfigboard.h,
> mpconfigboard.mk, or similar files — but I'm not sure where to start or
> how to properly define everything.
>
> I would be extremely grateful for any advice, examples, or working
> configurations!
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/micropython/discussions/16159#discussioncomment-12742335>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AK4TXJ6JCNES3LLXBFTRKSL2YFN2TAVCNFSM6AAAAABRGD6TBKVHI2DSMVQWIX3LMV43URDJONRXK43TNFXW4Q3PNVWWK3TUHMYTENZUGIZTGNI>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***
> com>
>


-- 
*André Botelho*
***@***.***

**Reply by AndreoBotelho** [reply]
Use CUBE IDE to help you choose the uarts and pinout, this MCU has 8 UARTs

**Comment by daleka** [comment]
[WEACT_H562RG.zip](https://github.com/user-attachments/files/19977730/WEACT_H562RG.zip)
Hello MicroPython community,

I'm building MicroPython firmware for the WeAct Studio STM32H562RGT6 board. The firmware compiles and uploads successfully via DFU, but after reset:

The USB (CDC) interface does not show up — the PC does not recognize the board.

Confirmed working:
Firmware builds successfully (.elf, .dfu, .hex)
DFU upload via dfu-util works fine
Board resets and starts running after upload

Not working:
USB is not detected by the system (not in dmesg, lsusb, or Windows Device Manager)
USB REPL (CDC) is not available
Connecting via USB cable does nothing

Already done:
In mpconfigboard.h, I have:
#define MICROPY_HW_ENABLE_USB               (1)
#define MICROPY_HW_USB_FS                         (1)
#define MICROPY_HW_USB_MAIN_DEV             (USB_PHY_FS_ID)
#define MICROPY_HW_USB_CDC                     (1)
#define MICROPY_HW_USB_MSC                     (1)

In pins.csv:
USB_DM,PA11
USB_DP,PA12

Other notes:
Tried adding/removing MICROPY_HW_CLK_USE_PLL3_FOR_USB
Verified power and USB connections (pull-up, caps, resistors)
Did a full make clean before building
USB works fine in a previous build (same board, different config)

Attached files:
mpconfigboard.h
mpconfigboard.mk
pins.csv
stm32h562rg.ld
stm32h5xx_hal_conf.h
board_init.c
bdev.c

🙏 I need help
Could someone take a quick look at my configuration files and help identify what’s missing or wrong? I'm stuck — USB works fine in another older build, but not in this cleaned-up version.

Thank you so much in advance for any insight!
[WEACT_H562RG.zip](https://github.com/user-attachments/files/19977734/WEACT_H562RG.zip)

**Reply by shariltumin** [reply]
I know; it was a while back. It's good if you manage to get it up and running. If not, I see a problem with your `stm32h562rg.ld` file.

```
  RAM (xrw)  : ORIGIN = 0x20000000, LENGTH = 640K

  _micropy_hw_internal_flash_storage_ram_cache_start = ORIGIN(RAM) + 640K - 8K;
  _micropy_hw_internal_flash_storage_ram_cache_end = ORIGIN(RAM) + 640K;

```

You will then have `flash_storage_ram_cache` overlap with your `RAM` and your board will not boot.

Changing these values as shown below may help resolve the issue:

```
   /* 640KB SRAM total, reserve last 8KB for FS cache  */
   RAM (xrw)      : ORIGIN = 0x20000000, LENGTH = 640K - 8K  /* 632KB */

 _micropy_hw_internal_flash_storage_ram_cache_start = ORIGIN(RAM) + LENGTH(RAM);
 _micropy_hw_internal_flash_storage_ram_cache_end   = ORIGIN(RAM) + 640;
```

**Reply by daleka** [reply]
Thank you for your reply and for taking the time to look into my issue. I actually sorted this problem out a long time ago, but I still really appreciate you reaching out and sharing your thoughts.
